### PR TITLE
[FEATURE] BERT Fine-tuning and Non-Deep Learning Approach

### PR DESCRIPTION
**Pull request by mrinoybanerjee**
# Implemented Bank Transaction categorization using a Fine-tuned BERT Model

## Description:
This pull request implements the first version of the feature to fine-tune a BERT model on categorized bank transaction data and an inference pipeline to add categories to unseen bank statements. In addition to the BERT model, this pull request also contains the non-deep learning approach for bank transaction categorization using TF-IDF and a Random Forest classifier.

## Changes Made:
Implemented first version of BERT model fine-tuning in a jupyter notebook script. Once tested, this notebook was refactored into a fully operational pipeline with a streamlit front-end. 

The pipeline consists of the following steps that have been separated out in scripts:

1. data_create.py: This script creates the data pipeline by taking in a bank statement csv file and checking if it has an existing category column. If it exists then it prepares the data for inference, otherwise uses BERT tokenizer to encode the data and creates data loaders to prepare for fine-tuning.
2. model.py: This Python script defines a custom BERT-based model for sequence classification tasks and includes functions for training, saving, loading, and predicting with the model using PyTorch and the Transformers library. The model architecture, training loop, and utility functions facilitate handling the classification task.
3. evaluate.py: This script provides a function to evaluate the BERT model using a PyTorch framework and Scikit-learn's metrics, calculating accuracy, precision, recall, and F1 score based on predictions from the dataloader. 
4. app.py: Invokes the streamlit frontend and calls the inference pipeline.

Finally this PR also contains the non-deep learning method notebook which trains a random forest classifier on the categorization task. Firstly class imbalance is handled by removing categories with lower than 10 transactions. Then TF-IDF is used to remove stop words from the descriptions and finally the RF classifier is trained and the hyperparameters are tuned using grid search.

## Screenshots
Screenshot 2024-04-12 at 2 58 35 PM

## Additional Notes:
For the next pull request need to add:

- Handling class imbalance before training the BERT model. 
- Creating a large synthetic transaction dataset to train the BERT classifier. Currently the dataset size might be too small with too many classes. 
- Hosting fine-tuned BERT model on hugging face and adding inference end points in the backend.

### CODE REVIEW

**Comment by mrinoybanerjee** [comment]
All checks passed, looks great and ready to be merged with main.
